rafthttp/stream.go

```diff
@@ -0,0 +1,279 @@
+package rafthttp
+
+import (
+	"encoding/binary"
+	"errors"
+	"fmt"
+	"io"
+	"log"
+	"net/http"
+	"net/url"
+	"path"
+	"sync"
+	"time"
+
+	"github.com/coreos/etcd/pkg/types"
+	"github.com/coreos/etcd/raft/raftpb"
+)
+
+var (
+	streamBufSize = 4096
+	errHeartbeat  = errors.New("the message is heartbeat")
+)
+
+type encoder interface {
```

Comment:
after looking at the code, i found the only necessary interface is encode. 

we should create a new encoder every time the "info" changed. also i do not think it is clear to let a encoder take a http.Header named as info and the function is actually called loadMetadata.